lib/srv/db/dynamodb/engine.go

```diff
@@ -0,0 +1,421 @@
+ limitations under the License.
+
+*/
+
+package dynamodb
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/service/dax"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/aws/aws-sdk-go/service/dynamodbstreams"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport"
+	apievents "github.com/gravitational/teleport/api/types/events"
+	apiaws "github.com/gravitational/teleport/api/utils/aws"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/teleport/lib/srv/db/common"
+	"github.com/gravitational/teleport/lib/srv/db/common/role"
+	"github.com/gravitational/teleport/lib/utils"
+	libaws "github.com/gravitational/teleport/lib/utils/aws"
+)
+
+func init() {
+	common.RegisterEngine(newEngine, defaults.ProtocolDynamoDB)
+}
+
+// newEngine create new DynamoDB engine.
+func newEngine(ec common.EngineConfig) common.Engine {
+	return &Engine{
+		EngineConfig:  ec,
+		RoundTrippers: make(map[string]http.RoundTripper),
+	}
+}
+
+// Engine handles connections from DynamoDB clients coming from Teleport
+// proxy over reverse tunnel.
+type Engine struct {
+	// signingSvc will be used by the engine to provide the AWS sigv4 authorization header
+	// required by AWS for request validation: https://docs.aws.amazon.com/general/latest/gr/signing-elements.html
+	signingSvc *libaws.SigningService
+	// EngineConfig is the common database engine configuration.
+	common.EngineConfig
+	// clientConn is a client connection.
+	clientConn net.Conn
+	// sessionCtx is current session context.
+	sessionCtx *common.Session
+	// RoundTrippers is a cache of RoundTrippers, mapped by service endpoint.
+	// It is not guarded by a mutex, since requests are processed serially.
+	RoundTrippers map[string]http.RoundTripper
+	// GetSigningCredsFn allows to set the function responsible for obtaining STS credentials.
+	// Used in tests to set static AWS credentials and skip API call.
+	GetSigningCredsFn libaws.GetSigningCredentialsFunc
+}
+
+var _ common.Engine = (*Engine)(nil)
+
+func (e *Engine) InitializeConnection(clientConn net.Conn, sessionCtx *common.Session) error {
+	e.clientConn = clientConn
+	e.sessionCtx = sessionCtx
+	awsSession, err := e.CloudClients.GetAWSSession(sessionCtx.Database.GetAWS().Region)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	svc, err := libaws.NewSigningService(libaws.SigningServiceConfig{
+		Clock:                 e.Clock,
+		Session:               awsSession,
+		GetSigningCredentials: e.GetSigningCredsFn,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	e.signingSvc = svc
+	return nil
+}
+
+// jsonErr is used to marshal a JSON error response as the AWS cli expects for errors.
+type jsonErr struct {
+	Code    string `json:"__type"`
+	Message string `json:"message"`
+}
```

Comment:
Could you add a reference to this format ?

Reply:
reference: https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Programming.Errors.html#Programming.Errors.Components

Reply:
although I found this originally by just digging deep into the aws sdk to see how they unmarshal errors. Before I added this the `aws` cli would print a useless generic error like "failed to unmarshal error response..." because it would try and fail to read the response error as json